source/base/OwnedBuffer.ooc

```diff
@@ -0,0 +1,73 @@
+/*
+* Copyright (C) 2015 - Simon Mika <simon@mika.se>
+*
+* This sofware is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+*
+* This software is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this software. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use ooc-base
+import math
+
+OwnedBuffer: cover {
+	_pointer: UInt8*
+	_size: Int
+	_owner: Owner
+	pointer ::= this _pointer
+	size ::= this _size
+	init: func@ {
+		this _pointer = null
+		this _size = 0
+		this _owner = Owner Unknown
+	}
+	init: func@ ~fromSize (size: Int) {
+		this init(gc_malloc(size), size, Owner Callee)
+	}
+	init: func@ ~fromData (=_pointer, =_size, =_owner)
+	take: func@ -> This {
+		this _changeOwnership(Owner Caller)
+		this
+	}
+	give: func -> This {
+		this _changeOwnership(Owner Callee) // call by value, works on copy
+		this
+	}
+	free: func@ -> Bool {
+		result: Bool
+		if ((result = (this _owner != Owner Literal && this _owner != Owner Stack && this _pointer != null))) {
+			gc_free(this _pointer)
+			this _pointer = null
+			this _size = 0
+		}
+		result
+	}
+	free: func@ ~withCriteria (criteria: Owner) -> Bool {
+		this _owner == criteria && this free()
+	}
+	copy: func -> This {
+		this _copyData()
+		this _owner = Owner Callee
+		this
```

Comment:
This looks really scary, but that's because I'm thinking of it as a class. Interesting stuff!